### ISSUE TITLE
[3.4] Variation selector breaks alignment

### ISSUE THREAD

**Issue by ncfavier** [tmux/tmux#3923]
### Issue description

After upgrading from tmux 3.3a to 3.4, weechat has started showing alignment/line length bugs around lines containing characters with a variation selector 16 (U+FE0F). I am using alacritty, which does not support these characters and instead just displays them on their own.

To reproduce, start weechat and run the following command:

```
/print 🅱️🅱️🅱️🅱️🅱️ test test test [...enough to spill onto the next line...] test test test
```

![](https://f.monade.li/PkYvA0.png)

That `test` on the left is not supposed to be there.

May be related to https://github.com/tmux/tmux/issues/3865, https://github.com/tmux/tmux/issues/3836#issuecomment-1962366253

The patch in https://github.com/tmux/tmux/issues/3865#issuecomment-1982019388 does *not* fix this issue.

### Required information

Please provide the following information:

* tmux version (`tmux -V`). 3.4
* Platform (`uname -sp`). Linux unknown
* $TERM inside and outside of tmux (`echo $TERM`). `alacritty`, `tmux-256color`
* Logs from tmux (`tmux kill-server; tmux -vv new`). https://gist.github.com/ncfavier/717ca38a48b5c873d0c2a447c0341180

**Comment by nicm** [comment]
tmux doesn't know your terminal doesn't support these, I expect we could add a capability and ignore the characters if it is not set but I'm not sure that would be any better.

**Comment by ncfavier** [comment]
Could you quickly explain what the problem is, and what changed in 3.4?

**Comment by nicm** [comment]
tmux now understands that VS are combining characters.

**Comment by ncfavier** [comment]
I can reproduce this in wezterm, which *does* support variation selectors.

![](https://f.monade.li/wdBnP-.png)

**Comment by nicm** [comment]
Does it still happen with master?

If so, can you come up with a `printf` command that shows the difference inside/outside tmux?

A variation selector on its own doesn't really matter that much, we should probably throw it out entirely.

**Comment by ncfavier** [comment]
Yes. `printf '|\U0001F171\uFE0F|\n'`.

**Comment by nicm** [comment]
\U and \u aren't portable and don't work on any of the platforms I have, can you show tmux logs from running this inside tmux?

**Comment by ncfavier** [comment]
Use the UTF-8 encoding then: `printf '|\xf0\x9f\x85\xb1\xef\xb8\x8f|\n'`. Logs: https://gist.github.com/ncfavier/fa33bb7d195402c569a6167e395d626f

**Comment by nicm** [comment]
OK I tried this with iTerm2, which does support VS, and the output is the same inside and outside tmux:
<img width="807" alt="Screenshot 2024-04-04 at 22 29 04" src="https://github.com/tmux/tmux/assets/304243/523441b9-dc79-4f4e-b29f-e216f7e789eb">
tmux is not sending anything different in your log from mine.
Can you show a screenshot of how it looks for you (in wezterm)?

**Comment by nicm** [comment]
This makes VS16 work with a previous non-wide-characters for me, at least with iTerm2, but we did this for a reason so I suspect it will break some characters (PUA?) which libc/utf8proc do not return as wide. Also it looks from your screenshot like wezterm renders it just as a space... so we maybe we can't predict what the terminal will do and would be better just to discard it, but then we have the same problem of figuring out if the previous character is suitable.
~~~~
diff --git a/screen-write.c b/screen-write.c
index 731c2cfd..a8e74807 100644
--- a/screen-write.c
+++ b/screen-write.c
@@ -2088,7 +2088,7 @@ screen_write_combine(struct screen_write_ctx *ctx, const struct grid_cell *gc)
        if (utf8_is_zwj(ud))
                zero_width = 1;
        else if (utf8_is_vs(ud))
-               zero_width = force_wide = 1;
+               zero_width = 1;
        else if (ud->width == 0)
                zero_width = 1;

**Comment by nicm** [comment]
Is the character two character positions wide in kitty? It looks like it is only one in wezterm.

If you make the change in this comment and rebuild tmux https://github.com/tmux/tmux/issues/3923#issuecomment-2038293073 does it fit tmux in wezterm? But break it in kitty?

**Comment by ncfavier** [comment]
Yes, two cells in kitty, one in wezterm. Your patch fixes wezterm, and doesn't seem to break kitty.

**Comment by nicm** [comment]
What if you do the `printf` inside tmux and then press `C-b r` (or `C-b : refresh-client`)? Does the text change in either terminal?

**Comment by nicm** [comment]
Funny, it doesn't break iTerm2 either but it really should because this means tmux is leaving the cell as width 1 (which is what utf8proc tells it) but the terminal is drawing it as width 2...

**Comment by ncfavier** [comment]
> What if you do the `printf` inside tmux and then press `C-b r` (or `C-b : refresh-client`)? Does the text change in either terminal?

Nope.

**Comment by nicm** [comment]
We get away with it drawing a pane somehow, but it is broken in other places if the terminal thinks it is width 2, for example if you do:
~~~~
tmux set -g status-left $(printf '|\xf0\x9f\x85\xb1\xef\xb8\x8f|')
~~~~
Then the status line is drawn incorrectly. And copy mode also positions the cursor incorrectly when moving across the character.

Unfortunately there is no way for tmux to know what width the terminal is actually going to use for any particular character. It would be nice if we could just trust libc or utf8proc to tell us - but in your case that would fix wezterm but break kitty. We chose to force these characters to width 2 because that is what most people and most terminals use, but it seems like wezterm or your font are different.

I think we need a way for people to override the width of individual characters or ranges of characters, then tmux could use the width it's told and if it's wrong people could just fix it themselves.

**Comment by nicm** [comment]
Here is another example it gets wrong with the above change on terminals where 🅱 is wide but wcwidth says it is not:
~~~~
printf 'X\xf0\x9f\x85\xb1\xef\xb8\x8fX';perl -e "print 'x'x(${COLUMNS}-3)";echo
~~~~
The last `x` should be on the next line (because the `X🅱X` should take up 4 cells) but it is not.

Anyway, I think we need to do something that allows the user to tell tmux what the terminal is going to do...

**Comment by ncfavier** [comment]
Why not just revert to the 3.3a behaviour?

**Comment by nicm** [comment]
Because it doesn't work for the same reasons I just gave..

**Comment by nicm** [comment]
You can try this which adds an option `variation-selector-always-wide` - if you turn if off it'll be right in wezterm, if you turn if on it'll be right in kitty: [x.diff.txt](https://github.com/tmux/tmux/files/14877917/x.diff.txt)

I'm not a big fan of this because it is a server option and it isn't very nice if you want to use both at the same time, or switch between them...

**Comment by ncfavier** [comment]
OK. Well, it's probably best to let the user decide if it's not clear what the terminals *ought* to do, but that's a bit of a shame. I certainly don't know though. Would be nice to have some sort of standard for displaying Unicode characters in a grid.

**Comment by rockorager** [comment]
@nicm All terminals which recognize VS selectors (at this time contour, foot, ghostty, and wezterm) will advertise support via a DECRQM response for mode 2027: https://github.com/contour-terminal/terminal-unicode-core. This might be helpful in determining how to handle the width of these since you can query at runtime.

**Comment by mitchellh** [comment]
> @nicm All terminals which recognize VS selectors (at this time contour, foot, ghostty, and wezterm) will advertise support via a DECRQM response for mode 2027: https://github.com/contour-terminal/terminal-unicode-core. This might be helpful in determining how to handle the width of these since you can query at runtime.

Mode 2027 is supposed to indicate a fully Unicode compliant terminal, and VS-awareness does not a fully compliant terminal make. It's probably still a stronger signal than nothing at all but terminals can be VS-aware without this. 

Sorry for dropping in, I'm the creator/author of [Ghostty](https://mitchellh.com/ghostty). It's still unreleased to the public, we have about 1000 testers with source access (@rockorager is one of them), and it'll be FOSS later this year. I'm only dropping in because Ghostty has full Unicode compliance and I regretfully know about this cursed subject. 😆 

I'm happy to see tmux gaining more capability on this. 😄 Although I'm going to guess a lot of other terminal emulators are going to be sad...

I admit I'm not a regular tmux user, but I took a look at `screen-write.c` and noticing some bugs. 

First:

https://github.com/tmux/tmux/blob/3c2621b41b9e1466d2b06718a0a8514f5d10ded9/screen-write.c#L2090-L2091

This assumes that VS16 automatically makes any preceding cell wide. I believe VS16 only applies to the emoji category of codepoints, so VS16 has to be paired with a UCD check. Otherwise, it does nothing. I know this code isn't portable, but for example `printf "xA\uFE0Fx"` should not make `A` wide, but in tmux 3.4 it does.

Second, I don't see any awareness of whether a codepoint is default-wide vs default-narrow. Admittedly, this is allowed by the spec to be "contextual" by the program implementing Unicode, so any answer is right. And unfortunately the Contour Unicode spec doesn't specify what the behavior is. For Ghostty, we use the default graphical vs text renditions specified in UCD (graphical = double width, text = single width). 

Sorry, I have to go, but those are a couple things I quickly noticed. I'll try to make a more in depth look at another time. 

❤️

**Comment by nicm** [comment]
I have applied the patch to add the option now (will be in GitHub later when it syncs), so you can set `variation-selector-always-wide off` if you are only using terminals which do not support this feature.

I looked at DECRQM 2027 but I couldn't get either iTerm2 or WezTerm to respond with anything useful that would let me detect their support (iTerm2 returns 4 and WezTerm returns 0, both before and after sending OSC 1337 UnicodeVersion=14). Anyway, this would require tmux to emulate the behaviour on terminals which do not support it, and I don't have either time to implement or interest in implementing that at the moment.
